calcom/cal.com
# Fix booking time hydration error

## What does this PR do?

Root cause of the error is that both server and client try to parse date differently due to localStorage. Since server does not know localStorage when render initial HTML it renders page as if there is no value set for time format.
 On the other hand, client side know the value of time format when it tries to render.

To fix the issue, we force it to parse date with default format at the initial rendering time.

- Add `withDefaultTimeFormat` argument to `parseDate` function
- Add `withDefaultTimeFormat` argument to `parseRecurringDates` function
- Leverage useEffect to dynamically update `withDefaultTimeFormat` arg passed to function

Fixes #7213 



**Environment**: Staging(main branch) 

## Type of change




## How should this be tested?

#### How to reproduce
- Create an event
- Go to book page by following `/[user]/[event]`
- Select any possible date and time after 13:00
- Make sure `timeOption.is24hClock` value on localStorage is `false`
- Refresh page or open new tab and try to access previous page
  - ex) `http://localhost:3000/teampro2/book?type=21&date=2023-02-23T23%3A45%3A00%2B09%3A00&slug=qwdqwd&count=`
- You will see the error

#### Check if it's fixed
- Just apply changes and try to reproduce

## Checklist



- I haven't checked if my changes generate no new warnings
- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes
